src/layers/ordered_leaf.rs

```diff
@@ -0,0 +1,397 @@
+//! Implementation using ordered keys and exponential search.
+
+use super::{advance, Builder, Cursor, MergeBuilder, Trie, TupleBuilder};
+use crate::{
+    algebra::{AddAssignByRef, AddByRef, HasZero, NegByRef},
+    NumEntries, SharedRef,
+};
+use std::{
+    marker::PhantomData,
+    ops::{Add, AddAssign, Neg},
+};
+
+/// A layer of unordered values.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct OrderedLeaf<K, R> {
+    /// Unordered values.
+    pub vals: Vec<(K, R)>,
+}
+
+// TODO: by-value merge
+impl<K, R> Add<Self> for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        self.merge(&rhs)
+    }
+}
+
+impl<K, R> AddAssign<Self> for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    fn add_assign(&mut self, rhs: Self) {
+        *self = self.merge(&rhs);
+    }
+}
+
+impl<K, R> AddAssignByRef for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    fn add_assign_by_ref(&mut self, other: &Self) {
+        *self = self.merge(other);
+    }
+}
+
+impl<K, R> AddByRef for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Eq + HasZero + AddAssignByRef + Clone,
+{
+    fn add_by_ref(&self, rhs: &Self) -> Self {
+        self.merge(rhs)
+    }
+}
+
+impl<K, R> NegByRef for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: NegByRef,
+{
+    fn neg_by_ref(&self) -> Self {
+        Self {
+            vals: self
+                .vals
+                .iter()
+                .map(|(k, v)| (k.clone(), v.neg_by_ref()))
+                .collect(),
+        }
+    }
+}
+
+impl<K, R> Neg for OrderedLeaf<K, R>
+where
+    K: Ord + Clone,
+    R: Neg<Output = R>,
+{
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        Self {
+            vals: self.vals.into_iter().map(|(k, v)| (k, v.neg())).collect(),
+        }
```

Comment:
This can probably be explicitly in-place
```suggestion
    fn neg(mut self) -> Self {
        for (_, value) in self.values.iter_mut() {
            *value = value.neg();
        }
    
        self
```